std/algebra/emulated/fields_bw6761/e6.go

```diff
@@ -105,10 +105,103 @@ func (e Ext6) Square(x *E6) *E6 {
 	}
 }
 
-// Karabina's compressed cyclotomic square
+// Karabina's compressed cyclotomic square SQR12345
+// https://eprint.iacr.org/2010/542.pdf
+// Sec. 5.6 with minor modifications to fit our tower
+func (e Ext6) CyclotomicSquareKarabina12345(x *E6) *E6 {
+	x = e.Reduce(x)
+
+	// h4 = -g4 + 3((g3+g5)(g1+c*g2)-g1g5-c*g3g2)
+	g1g5 := e.fp.Mul(&x.B0.A1, &x.B1.A2)
+	g3g2 := e.fp.Mul(&x.B1.A0, &x.B0.A2)
+	h4 := mulFpByNonResidue(e.fp, &x.B0.A2)
+	h4 = e.fp.Add(h4, &x.B0.A1)
+	t := e.fp.Add(&x.B1.A0, &x.B1.A2)
+	h4 = e.fp.Mul(h4, t)
+	h4 = e.fp.Sub(h4, g1g5)
+	t = mulFpByNonResidue(e.fp, g3g2)
+	h4 = e.fp.Sub(h4, t)
+	h4 = e.fp.MulConst(h4, big.NewInt(3))
+	h4 = e.fp.Sub(h4, &x.B1.A1)
+
+	// h3 = 2(g3+3c*g1g5)
+	h3 := mulFpByNonResidue(e.fp, g1g5)
+	h3 = e.fp.MulConst(h3, big.NewInt(3))
+	h3 = e.fp.Add(h3, &x.B1.A0)
+	h3 = e.fp.MulConst(h3, big.NewInt(2))
+
+	// h2 = 3((g1+g5)(g1+c*g5)-c*g1g5-g1g5)-2g2
+	t = mulFpByNonResidue(e.fp, &x.B1.A2)
+	t = e.fp.Add(t, &x.B0.A1)
+	h2 := e.fp.Add(&x.B1.A2, &x.B0.A1)
+	h2 = e.fp.Mul(h2, t)
+	h2 = e.fp.Sub(h2, g1g5)
+	t = mulFpByNonResidue(e.fp, g1g5)
```

Comment:
mulFpByNonResiude seems multiplication by -4 -- cannot we inline `-c*g1g5 - g1g5` as `-(c+1)g1g5`?

Reply:
absolutely :)